javalib/src/main/scala/java/lang/process/WindowsProcess.scala

```diff
@@ -0,0 +1,338 @@
+    def hasValidTimeout = timeout > 0L
+    def hasFinished =
+      WaitForSingleObject(
+        handle,
+        unit.toMillis(timeout).toUInt
+      ) match {
+        case WAIT_TIMEOUT => false
+        case WAIT_FAILED =>
+          throw WindowsException("Failed to wait on proces handle")
+        case _ => true
+      }
+
+    !isAlive() ||
+    (hasValidTimeout && hasFinished)
+  }
+
+  private[this] val _inputStream =
+    PipeIO[PipeIO.Stream](this, outHandle, builder.redirectOutput())
+  private[this] val _errorStream =
+    if (builder.redirectErrorStream()) PipeIO.InputPipeIO.nullStream
+    else PipeIO[PipeIO.Stream](this, errHandle, builder.redirectError())
+  private[this] val _outputStream =
+    PipeIO[OutputStream](this, inHandle, builder.redirectInput())
+
+  private def checkExitValue: Option[scala.Int] = {
+    checkResult()
+    cachedExitValue
+  }
+
+  private[lang] def checkResult(): CInt = {
+    cachedExitValue
+      .getOrElse {
+        val exitCode: Ptr[DWord] = stackalloc[DWord]
+        if (!GetExitCodeProcess(handle, exitCode)) -1
+        else {
+          (!exitCode) match {
+            case STILL_ACTIVE => -1
+            case code =>
+              _inputStream.drain()
+              _errorStream.drain()
+              _outputStream.close()
+              CloseHandle(handle)
+              cachedExitValue = Some(code.toInt)
+              code.toInt
+          }
+        }
+      }
+  }
+}
+
+object WindowsProcess {
+  type PipeHandles = CArray[Handle, Nat._2]
+  private final val readEnd = 0
+  private final val writeEnd = 1
+
+  private def zeroMemory[T: Tag](ptr: Ptr[T]) = {
+    import scalanative.libc.string.memset
+    memset(ptr.asInstanceOf[Ptr[Byte]], 0, sizeof[T])
+  }
+
+  def apply(builder: ProcessBuilder): Process = Zone { implicit z =>
+    val (inRead, inWrite) =
+      createPipeOrThrow(
+        builder.redirectInput(),
+        ConsoleApiExt.stdIn,
+        isStdIn = true,
+        "Couldn't create std input pipe."
+      )
+    val (outRead, outWrite) =
+      createPipeOrThrow(
+        builder.redirectOutput(),
+        ConsoleApiExt.stdOut,
+        isStdIn = false,
+        "Couldn't create std output pipe."
+      )
+    val (errRead, errWrite) = {
+      if (builder.redirectErrorStream()) (outRead, outWrite)
+      else
+        createPipeOrThrow(
+          builder.redirectError(),
+          ConsoleApiExt.stdErr,
+          isStdIn = false,
+          "Couldn't create std error pipe."
+        )
+    }
+
+    val cmd = builder.command().scalaOps.toSeq
+    val dir = toCWideStringUTF16LE(builder.directory().getAbsolutePath())
+    val argv = toCWideStringUTF16LE(cmd.mkString(" "))
+    val envp = nullTerminatedBlock {
+      builder
+        .environment()
+        .entrySet()
+        .scalaOps
+        .toSeq
+        .map(e => s"${e.getKey()}=${e.getValue()}")
+    }.asInstanceOf[Ptr[Byte]]
+
+    val startupInfo = stackalloc[StartupInfoW]
+    zeroMemory(startupInfo)
```

Comment:
Doesn't `stackalloc` already guarantee to return zeroed memory?

Reply:
Yes, that is correct, I've forgotten about that. Explicit zeroing of memory is redundant.